### PR TITLE
Do not default to suspending a job whose parent is already managed by Kueue

### LINKED ISSUE #1844: RayClusters managed by RayJobs stay in suspended state
**What happened**:

RayClusters that are created when RayJobs are admitted stay in suspended state, while they should be unsuspended and running. 

This has been first reported in https://github.com/kubernetes-sigs/kueue/pull/1802#issuecomment-1995580722.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a RayJob that can be admitted
2. Check the child RayCluster stays in suspended state

**Anything else we need to know?**:

**Environment**:
- Kueue version (use `git describe --tags --dirty --always`): v0.6.0-rc.1-116-ga96927de-dirty

### PR DESCRIPTION
**Pull request by astefanutti**
#### What type of PR is this?

/kind regression

#### Which issue(s) this PR fixes:

Fixes #1844

This is a regression introduced in #1747.

#### Does this PR introduce a user-facing change?

```release-note
Do not default to suspending a job whose parent is already managed by Kueue
```

### CODE REVIEW

**Comment by tenzen-y** [comment]
@astefanutti Could you update a release note?

**Comment by astefanutti** [comment]
@tenzen-y I wasn't sure what's your convention here. I think that's a regression that has only been introduced in main at the moment, is it still relevant to mention it in the release notes?

**Comment by tenzen-y** [comment]
> @tenzen-y I wasn't sure what's your convention here. I think that's a regression that has only been introduced in main at the moment, is it still relevant to mention it in the release notes?

Oh, I didn't find that. So, I'm ok with not having any release notes. Thanks!

File: pkg/controller/jobs/raycluster/raycluster_webhook.go [hunk 1]
```diff
@@ -60,6 +61,12 @@ func (w *RayClusterWebhook) Default(ctx context.Context, obj runtime.Object) err
 	log := ctrl.LoggerFrom(ctx).WithName("raycluster-webhook")
 	log.V(10).Info("Applying defaults", "job", klog.KObj(job))
 
+	// Do not default suspend a RayCluster that's managed by a RayJob
+	if owner := metav1.GetControllerOf(job.Object()); owner != nil && jobframework.IsOwnerManagedByKueue(owner) {
```
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_webhook.go:65]: Isn't this a problem for jobs too?

Comment (alculquicondor) [reply pkg/controller/jobs/raycluster/raycluster_webhook.go:65]: jobs that belong to jobsets, that is.

Comment (alculquicondor) [reply pkg/controller/jobs/raycluster/raycluster_webhook.go:65]: In that case, can we move the logic inside `jobframework.ApplyDefaultForSuspend`?

Comment (astefanutti) [reply pkg/controller/jobs/raycluster/raycluster_webhook.go:65]: That's a good idea. It'll apply to a larger set of types that's what required in practice , but I don't see any issue with that.

**Comment by astefanutti** [comment]
/retitle Do not default to suspending a job whose parent is already managed by Kueue

File: pkg/controller/jobs/raycluster/raycluster_webhook.go (hunk 1 above)
Comment (astefanutti) [reply pkg/controller/jobs/raycluster/raycluster_webhook.go]: Updated, PTAL.

File: test/integration/controller/jobs/raycluster/suite_test.go [hunk 2]
```diff
@@ -31,10 +31,12 @@ import (
 	"sigs.k8s.io/kueue/pkg/controller/core/indexer"
 	"sigs.k8s.io/kueue/pkg/controller/jobframework"
 	"sigs.k8s.io/kueue/pkg/controller/jobs/raycluster"
+	"sigs.k8s.io/kueue/pkg/controller/jobs/rayjob"
 	"sigs.k8s.io/kueue/pkg/queue"
 	"sigs.k8s.io/kueue/pkg/scheduler"
+	"sigs.k8s.io/kueue/pkg/webhooks"
 	"sigs.k8s.io/kueue/test/integration/framework"
-	//+kubebuilder:scaffold:imports
+	// +kubebuilder:scaffold:imports
```
Comment (alculquicondor) [inline test/integration/controller/jobs/raycluster/suite_test.go:39]: I don't think we need this line.

**Review by alculquicondor — COMMENTED** [review]
/approve

File: test/integration/controller/jobs/raycluster/suite_test.go (hunk 2 above)
Comment (astefanutti) [reply test/integration/controller/jobs/raycluster/suite_test.go]: that's right. Removed.

**Comment by alculquicondor** [comment]
@tenzen-y anything to add?

File: test/integration/controller/jobs/raycluster/raycluster_webhook_test.go [hunk 3]
```diff
@@ -56,10 +66,95 @@ var _ = ginkgo.Describe("RayCluster Webhook", func() {
...
+			gomega.Expect(k8sClient.Create(ctx, parentJob)).Should(gomega.Succeed())
+
+			lookupKey := types.NamespacedName{Name: parentJob.Name, Namespace: ns.Name}
+			gomega.EventuallyWithOffset(1, func() error {
+				if err := k8sClient.Get(ctx, lookupKey, parentJob); err != nil {
+					return err
+				}
+				parentJob.Status.JobDeploymentStatus = rayv1alpha1.JobDeploymentStatusSuspended
+				return k8sClient.Status().Update(ctx, parentJob)
+			}, util.Timeout, util.Interval).Should(gomega.Succeed())
```
Comment (tenzen-y) [inline test/integration/controller/jobs/raycluster/raycluster_webhook_test.go:118]:
```suggestion
```suggestion
			gomega.Eventually(func(g gomega.Gomega) {
				g.Expect(k8sClient.Get(ctx, lookupKey, parentJob)).Should(gomega.Succeeded())
				parentJob.Status.JobDeploymentStatus = rayv1alpha1.JobDeploymentStatusSuspended
				g.Expect(k8sClient.Status().Update(ctx, parentJob)).Should(gomega.Succeeded())
			}, util.Timeout, util.Interval).Should(gomega.Succeed())
```

Any offset isn't needed, right?

File: test/integration/controller/jobs/raycluster/raycluster_webhook_test.go [hunk 4]
```diff
@@ -56,10 +66,95 @@ var _ = ginkgo.Describe("RayCluster Webhook", func() {
...
+				Obj()
+			gomega.Expect(ctrl.SetControllerReference(parentJob, childCluster, k8sClient.Scheme())).To(gomega.Succeed())
+			gomega.Expect(k8sClient.Create(ctx, childCluster)).Should(gomega.Succeed())
+
+			childClusterKey := client.ObjectKeyFromObject(childCluster)
+			ginkgo.By("Checking that the child cluster is not suspended")
+			gomega.Eventually(func() *bool {
+				gomega.Expect(k8sClient.Get(ctx, childClusterKey, childCluster)).Should(gomega.Succeed())
+				return childCluster.Spec.Suspend
+			}, util.Timeout, util.Interval).Should(gomega.Equal(ptr.To(false)))
```
Comment (tenzen-y) [inline test/integration/controller/jobs/raycluster/raycluster_webhook_test.go:157]:
```suggestion
			gomega.Eventually(func(g gomega.Gomega) {
				gomega.Expect(k8sClient.Get(ctx, childClusterKey, childCluster)).Should(gomega.Succeed())
				gomega.Expect(childCluster.Spec.Suspend).Should(gomega.BeFalse())
			}, util.Timeout, util.Interval).Should(gomega.Succeeded)
```

**Review by tenzen-y — COMMENTED** [review]
Generally, lgtm

This error could happen in the case of MPIJob (MPIJob uses batch/v1 Job) as well.
Thank you!

File: test/integration/controller/jobs/raycluster/raycluster_webhook_test.go (hunk 3 above)
Comment (astefanutti) [reply test/integration/controller/jobs/raycluster/raycluster_webhook_test.go]: Right, updated in e2cee02b33adec66e756c9a5640ca6b2276a9152.

File: test/integration/controller/jobs/raycluster/raycluster_webhook_test.go (hunk 4 above)
Comment (astefanutti) [reply test/integration/controller/jobs/raycluster/raycluster_webhook_test.go]: Looks much better, updated in e2cee02b33adec66e756c9a5640ca6b2276a9152, thanks.

**Review by tenzen-y — COMMENTED** [review]
Thank you!
/lgtm
/approve

**Pull request merged**